### PR TITLE
Add custom hook types

### LINKED ISSUE #481: Hooks typescript definitions are missing
The typescript definitions for all the custom hooks are missing.

gz#5207

### PR DESCRIPTION
**Pull request by vicnicius**
## Description of the pull request

Moving some custom hook type definitions to `.d.ts`. Due to https://github.com/microsoft/TypeScript/issues/38691 it wasn't possible to do so on the version of TS we were using, so I upgraded TS and locked it to their current release candidate.

closes #481

### CODE REVIEW

File: src/components/MessageInput/hooks/messageInput.js [hunk 1]
```diff
@@ -208,7 +208,7 @@ function messageInputReducer(state, action) {
 }
 /**
  * hook for MessageInput state
- * @param {Props} props
+ * @type{import('types').useMessageInput}
```
Comment (mahboubii) [inline src/components/MessageInput/hooks/messageInput.js:211]:
```suggestion
 * @type{import('types').useMessageInputState}
```

File: src/components/MessageInput/hooks/messageInput.js [hunk 2]
```diff
@@ -665,6 +665,7 @@ export default function useMessageInputState(props) {
   );
 
   const onPaste = useCallback(
+    /** (event: React.ClipboardEvent) */
```
Comment (mahboubii) [inline src/components/MessageInput/hooks/messageInput.js:668]:
```suggestion
    /** (e: React.ClipboardEvent) */
```

File: types/index.d.ts [hunk 3]
```diff
@@ -694,6 +711,34 @@ export interface MessageInputEmojiPickerProps extends MessageInputState {
...
+    | undefined
+  )[];
+  getCommands(): Client.CommandResponse[] | undefined;
+  handleSubmit(event: React.FormEvent | React.MouseEvent): void;
+  handleChange(event: React.ChangeEventHandler): void;
+  onPaste(event: React.ClipboardEvent): void;
+  onSelectItem(item: Client.UserResponse): void;
+  openEmojiPicker(): void;
+}
+export function useMessageInput(
```
Comment (mahboubii) [inline types/index.d.ts:738]:
```suggestion
export function useMessageInputState(
```

**Review by mahboubii — APPROVED** [review]

**Comment by mahboubii** [comment]
closes #481
@vicnicius

File: src/components/MessageInput/hooks/messageInput.js (hunk 1 above)
Comment (vicnicius) [reply src/components/MessageInput/hooks/messageInput.js:211]: I started like that but noticed we export it named `useMessageInput` here: https://github.com/GetStream/stream-chat-react/blob/316cf2958e4fa81114beccd8a4753758404c4081/src/components/MessageInput/hooks/index.js

Comment (mahboubii) [reply src/components/MessageInput/hooks/messageInput.js:211]: oh right, should we rename it to `useMessageInput` so it's more consistent?

Comment (vicnicius) [reply src/components/MessageInput/hooks/messageInput.js:211]: 👍  Did it: https://github.com/GetStream/stream-chat-react/pull/601/commits/c3f981ac2563d2a30dde9903fded33cc524e311e

Comment (mahboubii) [reply src/components/MessageInput/hooks/messageInput.js:211]: Nice

**Review by mahboubii — APPROVED** [review]

**Pull request merged**